### ISSUE TITLE
Nesting of converters causes unexpected Exception

### ISSUE THREAD

**Issue by strindberg** [jOOQ/jOOQ#12525]
A MCVE for this issue can be found here: https://github.com/strindberg/jOOQ-mcve, branch "jooq_issue_12525"

Consider the following schema:

```sql
CREATE TABLE "user"
(
    id   bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    name text   NOT NULL,

    PRIMARY KEY (id)
);

CREATE TABLE event
(
    id                 bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    created_by_user_id bigint NOT NULL,

    PRIMARY KEY (id),
    FOREIGN KEY (created_by_user_id) REFERENCES "user" (id)
);
```

The following Kotlin model:

```kotlin
data class User(val id: Long? = null, val name: String)

data class Event(val id: Long? = null, val details: Details)

data class Details(val createdBy: User)
```

And the following select-method:

```kotlin
    fun getEventById(id: Long): Event? =
        dsl.select(
            EVENT.ID,
            field(
                row(
                    field(
                        row(
                            EVENT.user().ID,
                            EVENT.user().NAME,
                        )
                    ).convertFrom { toUser(it) }
                )
            ).convertFrom { toDetails(it) }
        ).from(EVENT).where(EVENT.ID.eq(id)).fetchOne { toEvent(it) }

    private fun toDetails(r: Record1<User>) = Details(r.value1()!!)

    private fun toEvent(r: Record2<Long?, Details>) = Event(r.value1()!!, r.value2())

    private fun toUser(r: Record2<Long?, String?>) = User(r.value1()!!, r.value2()!!)
```

Inserting an event and then running the select results in the following exception:

```
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64) ~[?:?]
	at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70) ~[?:?]
	at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248) ~[?:?]
	at java.util.Objects.checkIndex(Objects.java:373) ~[?:?]
	at java.util.ArrayList.get(ArrayList.java:427) ~[?:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.lambda$pgNewRecord$0(DefaultBinding.java:3745) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgNewRecord(DefaultBinding.java:3741) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3696) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3704) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3704) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgSetValue(DefaultBinding.java:3752) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.lambda$pgNewRecord$0(DefaultBinding.java:3745) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgNewRecord(DefaultBinding.java:3741) ~[jooq-3.15.3.jar:?]
	at org.jooq.impl.RowField.lambda$new$0(RowField.java:125) ~[jooq-3.15.3.jar:?]
	at org.jooq.Converters.lambda$nullable$0(Converters.java:213) ~[jooq-3.15.3.jar:?]
	at org.jooq.Converter$1.from(Converter.java:178) ~[jooq-3.15.3.jar:?]
...
```

### Versions

- jOOQ: 3.15.4

**Comment by strindberg** [comment]
I have upgraded the MCVE to use jOOQ 3.16.1. This release contained so many great bug fixes that I thought I might be lucky, but the problem remains.

**Comment by lukaseder** [comment]
Thanks for the ping. Looking into this now

**Comment by lukaseder** [comment]
Ah, yes. It seems that `field(row())` has simply not yet been supported, an oversight when https://github.com/jOOQ/jOOQ/issues/11812 was implemented. A backport to 3.16 and 3.15 would be advised, then, if this can be fixed.

**Comment by lukaseder** [comment]
Fix seems trivial. The internal `AbstractRow.rf()` auxiliary method already provides the correct implementation. With the fix, your test case now works:

```
14:18:36,855 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : delete from "public"."event"
14:18:36,869 DEBUG [org.jooq.tools.LoggerListener                     ] - Affected row(s)          : 1
14:18:36,869 DEBUG [org.jooq.impl.AbstractQuery                       ] - DELETE without WHERE     : A statement is executed without WHERE clause
14:18:36,870 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : delete from "public"."user"
14:18:36,874 DEBUG [org.jooq.tools.LoggerListener                     ] - Affected row(s)          : 1
14:18:36,927 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : insert into "public"."user" ("name") values (?) returning "public"."user"."id"
14:18:36,928 DEBUG [org.jooq.tools.LoggerListener                     ] - -> with bind values      : insert into "public"."user" ("name") values ('Name') returning "public"."user"."id"
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched result           : +----+
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |  id|
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |   4|
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+
14:18:36,957 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched row(s)           : 1
14:18:36,960 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : insert into "public"."event" ("created_by_user_id") values (?) returning "public"."event"."id"
14:18:36,961 DEBUG [org.jooq.tools.LoggerListener                     ] - -> with bind values      : insert into "public"."event" ("created_by_user_id") values (4) returning "public"."event"."id"
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched result           : +----+
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |  id|
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |   4|
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+
14:18:36,966 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched row(s)           : 1
14:18:37,209 DEBUG [org.jooq.tools.LoggerListener                     ] - Executing query          : select "public"."event"."id", row (row ("alias_90521876"."id", "alias_90521876"."name")) from ("public"."event" join "public"."user" as "alias_90521876" on "public"."event"."created_by_user_id" = "alias_90521876"."id") where "public"."event"."id" = ? -- SQL rendered with a free trial version of jOOQ 3.17.0-SNAPSHOT
14:18:37,210 DEBUG [org.jooq.tools.LoggerListener                     ] - -> with bind values      : select "public"."event"."id", row (row ("alias_90521876"."id", "alias_90521876"."name")) from ("public"."event" join "public"."user" as "alias_90521876" on "public"."event"."created_by_user_id" = "alias_90521876"."id") where "public"."event"."id" = 4
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched result           : +----+----------------------------------------+
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |  id|row                                     |
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+----------------------------------------+
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : |   4|Details(createdBy=User(id=4, name=Name))|
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] -                          : +----+----------------------------------------+
14:18:37,219 DEBUG [org.jooq.tools.LoggerListener                     ] - Fetched row(s)           : 1
Event(id=4, details=Details(createdBy=User(id=4, name=Name)))
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.295 sec

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.2 (https://github.com/jOOQ/jOOQ/issues/12870), and 3.15.6 (https://github.com/jOOQ/jOOQ/issues/12871)

**Comment by lukaseder** [comment]
I'm releasing 3.16.2 today due to a significant regression for H2 users (https://github.com/jOOQ/jOOQ/issues/12864), so you can use this fix soon.
